### PR TITLE
Refactor Deserialization

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring

## :scroll: Description

Use single method to deserialize JSON.

## :bulb: Motivation and Context

https://github.com/getsentry/sentry-java/pull/971#discussion_r523730386

## :green_heart: How did you test it?

Unit tests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [x] All tests passing
- [ ] No breaking changes

### CODE REVIEW

File: sentry/src/main/java/io/sentry/GsonSerializer.java [hunk 1]
```diff
@@ -102,49 +102,17 @@ public GsonSerializer(
...
+   * Deserialize an object of class given by {@code clazz} parameter from a stream Reader (JSON)
    *
    * @param reader the Reader
-   * @return the SentryEvent class or null
+   * @param clazz the type of object to deserialize into
+   * @return the deserialized object or null
    */
   @Override
-  public @Nullable SentryEvent deserializeEvent(final @NotNull Reader reader) {
+  public <T> T deserialize(Reader reader, Class<T> clazz) {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/GsonSerializer.java:112]:
```suggestion
  public <T> T deserialize(final @NotNull Reader reader, final @NotNull Class<T> clazz) {
```

File: sentry/src/main/java/io/sentry/GsonSerializer.java [hunk 2]
```diff
@@ -102,49 +102,17 @@ public GsonSerializer(
...
     Objects.requireNonNull(reader, "The Reader object is required.");
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/GsonSerializer.java:113]:
```suggestion
    Objects.requireNonNull(reader, "The Reader object is required.");
    Objects.requireNonNull(clazz, "The Class type is required.");
```

**Comment by marandaneto** [comment]
looks good, missing changelog though.

**Review by marandaneto — APPROVED** [review]
LGTM

**Comment by maciejwalkowiak** [comment]
@bruno-garcia perhaps you can also take a look at this one since @marandaneto asked to make it a part of the Performance PR too.

**Review by bruno-garcia — APPROVED** [review]

**Pull request merged**